### ISSUE TITLE
Unable to open SPI device: No such file or directory

### ISSUE THREAD

**Issue by demenkovms** [nopnop2002/Raspberry-ili9340spi#28]
Try to use it on Orange Pi Zero with WiringPi.

I connect TFT to OPI zero, to SPI 1 

```
TFT                                OPI
SDO(MISO)  -------->   SPI1_MISO/PA16
SCK             -------->    SPI1_CLK/PA14
SDI(MOSI)    -------->   SPI1_MOSI/PA15
DC               -------->   PA12
RESET           -------->   PA11
CS                -------->   PA13
VCC, LED     -------->    3.3v
GND            -------->    GND

set in ili9340.c
#define D_C  12
#define RES  11
#define C_S  13
```

make 
`cc -o demo demo.c fontx.c ili9340.c -lwiringPi -lm -DWPI`

run ./demo and get:

```
Your TFT resolution is 240 x 320.
Your TFT offsetx    is 0.
Your TFT offsety    is 0.
Opening device /dev/spidev0.0
Unable to open SPI device: No such file or directory
```

Please, help: what I did wrong? Thanks!

**Comment by nopnop2002** [comment]
>Opening device /dev/spidev0.0

Orange Pi Zero's CS is CS1, not CS0.

**Comment by demenkovms** [comment]
Thanks for answer! Not I get 
![изображение](https://user-images.githubusercontent.com/89016279/229289285-f03bafe8-2f90-4dd3-9eec-e975583fca7e.png)

but nothing shows on display - it's white

**Comment by nopnop2002** [comment]
```
set in ili9340.c
#define D_C  12
#define RES  11
#define C_S  13
```
According to this, OrangePi ZERO does not have GPIO11,12,13.

https://esp8266-server.de/OrangePi/

**Comment by demenkovms** [comment]
hm... 
in my 
![изображение](https://user-images.githubusercontent.com/89016279/229306054-4b40b919-1294-4c71-89b3-f0ef3ac2ba45.png)

is I need set number by from GPIO or wPi?

**Comment by nopnop2002** [comment]
This project uses the wiringPiSetupGpio() function to initialize GPIOs.

You need to check the GPIO pinmap.

Does the following command blink the LED connected to pin number 8?

```
$ gpio -g mode 198 out
$ gpio -g write 198 on
$ gpio -g write 198 of
```

**Comment by demenkovms** [comment]
led is not not blink and gpio not made on...

![изображение](https://user-images.githubusercontent.com/89016279/229352620-0b00652f-f0f9-4e94-bbbf-284acf0d8b1e.png)

**Comment by demenkovms** [comment]
I download this library https://github.com/xpertsavenue/WiringOP-Zero and it works correct with LED.

![изображение](https://user-images.githubusercontent.com/89016279/229353691-dc4b8fee-b8f4-4052-950f-33eb53723305.png)

then I change in ili9340.c pins from wPi

`#define D_C  8 
#define RES  9 
#define C_S  10
`
recompiled demo, but nothing changed...

**Comment by demenkovms** [comment]
I download this library https://github.com/xpertsavenue/WiringOP-Zero and it works correct with LED.

![изображение](https://user-images.githubusercontent.com/89016279/229353691-dc4b8fee-b8f4-4052-950f-33eb53723305.png)

then I change in ili9340.c pins from wPi

`#define D_C  8 
#define RES  9 
#define C_S  10
`
recompiled demo, but nothing changed...

**Comment by nopnop2002** [comment]
This lib is for legacy kernel.

>SPI is tested on Armbian with legacy kernel:

**Comment by demenkovms** [comment]
I need to downgrade kernel to 5.10. Is I correctly understood?

**Comment by nopnop2002** [comment]
You are correct, but there is no mention of OPI-ZERO anywhere.

OPI-ZERO is SPI1, so there may be other issues.

**Comment by demenkovms** [comment]
then, may be it’s possible to change spi in library code?

**Comment by nopnop2002** [comment]
I can’t understand your mention.

**Comment by demenkovms** [comment]
I mean: in your library used /dev/spi0, and in Opi Zero used /dev/spi1. Can I change spi0 to spi1 in code?

**Comment by nopnop2002** [comment]
>Can I change spi0 to spi1 in code?

Use compile flag.
-DSPI1

**Comment by demenkovms** [comment]
yes. I compiled with this flag, as you wrote some posts earlier. It’s runs correctly but nothing shows on display

**Comment by nopnop2002** [comment]
Using the latest repo, I have verified that it works fine on OrangePiPC.

```
$ uname -a
Linux orangepipc 5.10.60-sunxi #21.08.1 SMP Wed Aug 25 18:19:32 UTC 2021 armv7l GNU/Linux
```

RES-->Pin#5
DC-->Pin#3

```
git clone https://github.com/nopnop2002/Raspberry-ili9340spi
cd Raspberry-ili9340spi
cc -o demo demo.c fontx.c ili9340.c -lwiringPi -lm -pthread -DWPI
sudo ./demo
```

![IMG_3601](https://user-images.githubusercontent.com/6020549/229385860-d40b19f8-fd5d-42a5-bb98-1dc75002bc1a.JPG)

Try latest, and with -DSPI1.

**Comment by nopnop2002** [comment]
OrangePi ONE works fine.

However, it doesn't work properly with OrangePi ZERO.

It's probably a SPI driver issue.

![IMG_3604](https://user-images.githubusercontent.com/6020549/229410062-38b96482-594a-4bad-a0d5-5a20c09a5cd2.JPG)

**Comment by nopnop2002** [comment]
param_spidev_spi_bus=0 ---> param_spidev_spi_bus=1


DC --------> PA12 (gpio 12 - wPi 0 - pin 3) --> #define D_C  0  // wPi IO0=Pin#3


RESET --------> PA11 (gpio 11 - wPi 1 - pin 5) --> #define RES  1  // wPi IO1=Pin#5

cc -o demo demo.c fontx.c ili9340.c -lwiringPi -lm -pthread -DWPI -DSPI1

**Comment by nopnop2002** [comment]
I tried this and it doesn't work.



```
$ uname -a
Linux orangepizero 5.10.34-sunxi #21.05.1 SMP Thu May 6 20:13:21 UTC 2021 armv7l GNU/Linux

$ cat /boot/armbianEnv.txt
verbosity=1
bootlogo=false
console=serial
disp_mode=1920x1080p60
overlay_prefix=sun8i-h3
overlays=usbhost2 usbhost3
rootdev=UUID=93502503-5a81-4c0c-b76e-fabcad07a82b
rootfstype=ext4
overlays=spi-spidev spi-add-cs1
param_spidev_spi_bus=1
param_spidev_spi_cs=0
usbstoragequirks=0x2537:0x1066:u,0x2537:0x1068:u
```

**Comment by nopnop2002** [comment]
Hardware SPI doesn't work properly on ZERO.

So we supported software SPI.
Very slow but work.

I used this library.

https://github.com/xpertsavenue/WiringOP-Zero

```
$ uname -a
Linux orangepizero 5.15.93-sunxi #23.02.2 SMP Fri Feb 17 23:49:46 UTC 2023 armv7l GNU/Linux
```

![IMG_3608](https://user-images.githubusercontent.com/6020549/230514416-7a9594f6-1053-4683-a597-5b47e1a89b12.JPG)
